src/main/java/io/warp10/ext/interpolation/InterpolatorOnTicks.java

```diff
@@ -0,0 +1,159 @@
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+//   Unless required by applicable law or agreed to in writing, software
+//   distributed under the License is distributed on an "AS IS" BASIS,
+//   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//   See the License for the specific language governing permissions and
+//   limitations under the License.
+//
+
+package io.warp10.ext.interpolation;
+
+import io.warp10.continuum.gts.GTSHelper;
+import io.warp10.continuum.gts.GeoTimeSerie;
+import io.warp10.script.NamedWarpScriptFunction;
+import io.warp10.script.WarpScriptException;
+import io.warp10.script.WarpScriptMapperFunction;
+import io.warp10.script.WarpScriptStack;
+import io.warp10.script.WarpScriptStackFunction;
+import io.warp10.script.functions.SNAPSHOT;
+import org.apache.commons.math3.analysis.interpolation.AkimaSplineInterpolator;
+import org.apache.commons.math3.analysis.interpolation.LinearInterpolator;
+import org.apache.commons.math3.analysis.interpolation.SplineInterpolator;
+import org.apache.commons.math3.analysis.polynomials.PolynomialSplineFunction;
+
+/**
+ * Function that implements 1D interpolator on ticks
+ */
+public class InterpolatorOnTicks extends NamedWarpScriptFunction implements WarpScriptStackFunction {
+
+  public enum TYPE {
+    LINEAR, SPLINE, AKIMA
+  }
+
+  private TYPE type;
+
+  private static class MAPPER_INTERPOLATOR extends NamedWarpScriptFunction implements WarpScriptMapperFunction {
+
+    private final PolynomialSplineFunction func;
+    private final String generatedFrom;
+    private GeoTimeSerie gts;
+
+    private MAPPER_INTERPOLATOR(PolynomialSplineFunction function, String interpolatorName, GeoTimeSerie gts) {
+      super(interpolatorName);
+      func = function;
+      this.gts = gts;
+      generatedFrom = interpolatorName;
+    }
+
+    private double value(double x) {
+      if (!func.isValidPoint(x)) {
+        return Double.NaN;
+      } else {
+        return func.value(x);
+      }
+    }
+
+    @Override
+    public Object apply(Object[] args) throws WarpScriptException {
+      long tick = (long) args[0];
+      long[] ticks = (long[]) args[3];
+      long[] locations = (long[]) args[4];
+      long[] elevations = (long[]) args[5];
+      Object[] values = (Object[]) args[6];
+
+      if (1 < ticks.length) {
+        throw new WarpScriptException(getName() + " expects at most 1 tick but got " + ticks.length);
+      }
+
+      //
+      // We return the same value if a value is present,
+      // else, we interpolate.
+      //
+
+      if (1 == values.length) {
+        double value = ((Number) values[0]).doubleValue();
+
+        if (!Double.isNaN(value)) {
+          return new Object[] {tick, locations[0], elevations[0], value};
+
+        } else {
+          return new Object[] {tick, locations[0], elevations[0], value(tick)};
+        }
+      }
+
+      return new Object[] {tick, locations[0], elevations[0], value(tick)};
```

Comment:
locations, elevations, can be size zero.

Reply:
by the way, why not train the interpolator on elevation and geo data too ?  (four interpolators in one)

Reply:
To leave more flexibility. That could be different interpolation models. Also, that's a lot of complexity for an edge case. It's better to not touch geo data but it can be interpolated by other means.

Reply:
bug is fixed